src/main/java/sirius/biz/process/ErrorContext.java

```diff
@@ -0,0 +1,217 @@
+     *
+     * @param label the name of the value (this will be {@link NLS#smartGet(String) auto translated}.
+     * @param value the value to store
+     * @return the extended context for fluent method calls
+     */
+    public ErrorContext withContext(String label, Object value) {
+        if (Strings.isEmpty(value)) {
+            this.context.remove(label);
+        } else {
+            this.context.put(label, NLS.toUserString(value));
+        }
+
+        return this;
+    }
+
+    /**
+     * Removes the value with the given label from the context.
+     *
+     * @param label the label of the value to remove
+     * @return the modified context for fluent method calls
+     */
+    public ErrorContext removeContext(String label) {
+        return withContext(label, null);
+    }
+
+    /**
+     * Adds the given value to the context, performs the given task and then removes the value again.
+     *
+     * @param label the name of the value (this will be {@link NLS#smartGet(String) auto translated}.
+     * @param value the value to store
+     * @param task  the task to perform while the value is set
+     * @return the context for fluent method calls
+     * @throws Exception any error as thrown by the given task. Note that the exceptions and not handlend and also
+     *                   not automatically {@link #enhanceMessage(String) enhanced}.
+     */
+    public ErrorContext inContext(String label, Object value, UnitOfWork task) throws Exception {
+        withContext(label, value);
+        task.execute();
+        removeContext(label);
+
+        return this;
+    }
+
+    /**
+     * Obtains the current context as string.
+     *
+     * @return the current error context as string
+     */
+    public String getContextAsString() {
+        return context.entrySet()
+                      .stream()
+                      .map(entry -> NLS.smartGet(entry.getKey()) + ": " + entry.getValue())
+                      .collect(Collectors.joining(", "));
+    }
+
+    /**
+     * Performs the given task and handles / {@link #enhanceMessage(String) enhances} all thrown errors.
+     *
+     * @param failureDescription annotates a given error message so that the user is notified what task actually went
+     *                           wrong. This should be in "negative form" like "Cannot perform x because: message" as
+     *                           it is only used for error reporting.
+     * @param task               the task to actually perform
+     */
+    public void perform(UnaryOperator<String> failureDescription, UnitOfWork task) {
+        try {
+            task.execute();
+        } catch (HandledException exception) {
+            if (exception.getHint(MESSAGE_ENHANCED).asBoolean()) {
+                throw exception;
+            }
+
+            throw Exceptions.createHandled()
+                            .error((exception.getCause() instanceof HandledException) ? null : exception.getCause())
```

Comment:
hm. Why not always add the catched exception as the cause?

Reply:
If it is handled, our newly created exception would be ignored and our context would be lost..